Fix to not allow parameter entity references at internal subsets

naitoh opened the pull request:
## Why?
In the internal subset of DTD, references to parameter entities are not allowed within markup declarations.

See: https://www.w3.org/TR/xml/#wfc-PEinInternalSubset

> Well-formedness constraint: PEs in Internal Subset

> In the internal DTD subset, parameter-entity references MUST NOT occur within markup declarations; they may occur where markup declarations can occur. (This does not apply to references that occur in external parameter entities or to the external subset.)

kou reviewed (commented):
Wow!

Can we also check this in `BaseParser`?
If we can do it, can we move the validation to `#write`?

Do we need to update `#unnormalized` and `#normalized` too?
https://github.com/ruby/rexml/blob/e3f747fb4fe30f5c890a4bea5b12dd72f595e6b3/lib/rexml/entity.rb#L70-L79
https://github.com/ruby/rexml/blob/e3f747fb4fe30f5c890a4bea5b12dd72f595e6b3/lib/rexml/entity.rb#L83-L87

naitoh commented on lib/rexml/entity.rb:
```diff
@@ -123,43 +123,14 @@ def to_s
-          end
-          rv.gsub!( /%#{entity_reference.join};/um, entity_value )
-        end
+      if @value.match?(PEREFERENCE_RE)
```
OK, I see.

naitoh commented:
> Wow!
> 
> Can we also check this in `BaseParser`?

I changed BaseParser to do this check. 

> If we can do it, can we move the validation to `#write`?

I added validation to Entity.new.

> Do we need to update `#unnormalized` and `#normalized` too?

Updated `#unnormalized`.
No need to update `#normalized`.

kou commented on lib/rexml/entity.rb:
```diff
@@ -59,6 +60,10 @@ def initialize stream, value=nil, parent=nil, reference=false
       end
+
+      if PEREFERENCE_RE.match?(@value)
+        raise "Parameter Entity References forbidden in internal subset: #{@value}"
```
* Could you use `ArgumentError` instead of `RuntimeError`?
* Could you avoid needless capital case?

```suggestion
        raise ArgumentError, "Parameter entity references forbidden in internal subset: #{@value}"
```

BTW, do we need a validation here? This method document says the following:

> +WARNING+: There is no validation of entity state except when the entity
> is read from a stream.  If you start poking around with the accessors,
> you can easily create a non-conformant Entity.

naitoh replied:
@kou 
OK.
I see.
We prefer not to add this validation in `.new`.

Similarly, I don't think this validation should be added to `#write`.
What do you think?

kou replied:
I agree with you.

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -125,6 +125,7 @@ class BaseParser
       }
 
       module Private
+        PEREFERENCE_RE = /#{PEREFERENCE}/um
```
```suggestion
        PEREFERENCE_PATTERN = /#{PEREFERENCE}/um
```

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -334,6 +335,8 @@ def pull_event
                 match.delete_at(5) if match.size > 5 # Chop out NDATA decl
                 # match is [ :entity, name, PUBLIC, pubid, href(, ndata)? ]
+              elsif Private::PEREFERENCE_RE.match?(match[2])
+                raise REXML::ParseException.new("Parameter Entity References forbidden in internal subset: #{match[2]}", @source)
```
```suggestion
                raise REXML::ParseException.new("Parameter entity references forbidden in internal subset: #{match[2]}", @source)
```

kou commented on test/test_entity.rb:
```diff
@@ -104,6 +102,62 @@ def test_replace_entities
       assert_equal source, out
     end
 
+    def test_entity_with_reference
```
Can we improve test name? This class is `EntityTester`. So `_entity` is redundant.

`test_readers_reference`?

kou commented on test/test_entity.rb:
```diff
@@ -104,6 +102,62 @@ def test_replace_entities
 
+    def test_entity_with_reference
+      entity = REXML::Entity.new( [:entitydecl, "a", "B", "%"])
+      assert_equal('<!ENTITY % a "B">', entity.to_s)
+      assert_equal("a", entity.name)
+      assert_equal("B", entity.value)
+      assert_equal("B", entity.normalized)
+      assert_equal("B", entity.unnormalized)
```
Could you check by one `assert_equal` for easy to debug?
FYI: https://www.clear-code.com/blog/2011/2/28.html

```suggestion
      assert_equal([
                     '<!ENTITY % a "B">',
                     "a",
                     "B",
                     "B",
                     "B",
                   ],
                   [
                     entity.to_s,
                     entity.name,
                     entity.value,
                     entity.normalized,
                     entity.unnormalized,
                   ])
```

kou commented:
Thanks.
